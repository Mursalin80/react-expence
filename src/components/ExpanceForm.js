import React, { Component } from "react";

export default class ExpanceForm extends Component {
  constructor(props) {
    super(props);

    this.state = {
      note: props.expence ? props.expence.note : "",
      createdAt: props.expence ? props.expence.createdAt : 0,
      description: props.expence ? props.expence.description : "",
      amount: props.expence ? props.expence.amount : 0,
      error: ""
    };
  }

  onChangeHandler = e => {
    let { name, value } = e.target;
    this.setState(() => ({ [name]: value }));
  };

  onSubmitHandler = e => {
    e.preventDefault();
    let { note, description, amount, createdAt } = this.state;
    if (!note && !amount) {
      this.setState({ error: "No note and amount is provided." });
    }
    this.props.onSubmit({
      note,
      description,
      amount,
      createdAt
    });
  };

  render() {
    return (
      <div className="ui container">
        <form onSubmit={this.onSubmitHandler} className="ui form">
          <input
            className="field"
            type="text"
            name="note"
            placeholder="note"
            value={this.state.note}
            onChange={this.onChangeHandler}
          />
          <input
            className="field"
            type="text"
            name="amount"
            placeholder="amount"
            value={this.state.amount}
            onChange={this.onChangeHandler}
          />
          <input
            className="field"
            type="text"
            name="description"
            placeholder="description"
            value={this.state.description}
            onChange={this.onChangeHandler}
          />
          <input
            className="field"
            type="text"
            name="createdAt"
            placeholder="createdAt"
            value={this.state.createdAt}
            onChange={this.onChangeHandler}
          />

          <button className="ui button">Add Expence</button>
        </form>
      </div>
    );
  }
}
