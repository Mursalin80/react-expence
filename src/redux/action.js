import { v4 as uuidv4 } from "uuid";

export const getVisibalExpances = (expances, { note = "" } = {}) => {
  return expances.filter(exp => exp.note.includes(note));
};

export const getVisExpSortBy = (
  expances,
  { note = "", sortBy = "date" } = {}
) => {
  return expances
    .filter(exp => exp.note.includes(note))
    .sort((a, b) => {
      if (sortBy === "date") {
        return a.createdAt > b.createdAt ? 1 : -1;
      } else if (sortBy === "amount") {
        return a.amount > b.amount ? 1 : -1;
      }
    });
};

// add expance action
export const addExpance = ({
  note = "",
  amount = 0,
  desc = "",
  createdAt = 0
} = {}) => ({
  type: "Add_exp",
  expence: { id: uuidv4(), note, amount, desc, createdAt }
});

// remove expance action
export const removeExpance = ({ id } = {}) => ({
  type: "Remove_exp",
  id
});

// Edit expance action
export const editExpance = ({ id, update }) => ({
  type: "Edit_exp",
  id,
  payload: update
});

// add note filter action

export const noteFilter = (note = "") => ({
  type: "Note_filter",
  note
});

// sortBy filter action

export const sortByDate = () => ({
  type: "Sort_by_Date"
});

export const sortByAmount = () => ({
  type: "Sort_by_Amount"
});
