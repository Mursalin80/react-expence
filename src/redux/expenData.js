export const expData = [
  {
    id: 1,
    note: "gas bill",
    description: "gas bill for month of Jan 2020",
    amount: 250,
    createdAt: 2000
  },
  {
    id: 2,
    note: "net bill",
    description: "net bill for month of Jan 2020",
    amount: 1000,
    createdAt: 3600
  },
  {
    id: 3,
    note: "Milk bill",
    description: "Milk bill for month of Jan 2020",
    amount: 5600,
    createdAt: 5300
  },
  {
    id: 4,
    note: "Milk bill feb",
    description: "Milk bill for month of Feb 2020",
    amount: 5809,
    createdAt: 6000
  },
  {
    id: 5,
    note: "Milk bill mar",
    description: "Milk bill for month of Mar 2020",
    amount: 4500,
    createdAt: 6500
  }
];
