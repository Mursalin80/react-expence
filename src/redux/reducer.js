const initState = [];
export const initFilterState = {
  note: "",
  amount: 0,
  createdAt: undefined,
  sortBy: ""
};

export const expenReducer = (state = initState, action) => {
  switch (action.type) {
    case "Add_exp":
      return [...state, action.expence];
    case "Remove_exp":
      return state.filter(({ id }) => id !== action.id);
    case "SET_exp":
      return action.expances;
    case "Edit_exp":
      return state.map(exp => {
        if (exp.id === action.id) {
          return { ...exp, ...action.payload };
        } else {
          return exp;
        }
      });
    default:
      return state;
  }
};

// filter Reduser
export const filterReducer = (state = initFilterState, action = {}) => {
  switch (action.type) {
    case "Note_filter":
      return { ...state, note: action.note };
    case "Sort_by_Date":
      return { ...state, sortBy: "date" };
    case "Sort_by_Amount":
      return { ...state, sortBy: "amount" };
    default:
      return state;
  }
};

export const authReducer = (state = {}, action) => {
  switch (action.type) {
    case "LOGIN":
      return {
        uid: action.user.uid,
        user: action.user
      };
    case "LOGOUT":
      return {};
    default:
      return state;
  }
};
