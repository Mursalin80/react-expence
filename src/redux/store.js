import { createStore, combineReducers, applyMiddleware } from "redux";
import thunk from "redux-thunk";
import { composeWithDevTools } from "redux-devtools-extension";
import { expenReducer, filterReducer } from "./reducer";
// import { expData } from "./expenData";
// import {
//   addExpance,
//   editExpance,
//   // getVisibalExpances,
//   noteFilter,
//   sortByAmount
// } from "./action";

console.log("Redux is running");

// Expence Reduser

// store
export let store = createStore(
  combineReducers({ expenReducer, filterReducer }),
  composeWithDevTools(applyMiddleware(thunk))
);

// store.subscribe(() => {
//   // console.log("store.subscribe: ", store.getState());
//   // const state = store.getState();
//   // let visibleExp = getVisibalExpances(state.expenReducer, state.filterReducer);
//   // console.log("state.filterReducer : ", state.filterReducer);
//   // console.log("Visible Expances: ", visibleExp);
// });

// store.dispatch(
//   addExpance({
//     note: "Milk bill",
//     Desc: "Milk bill for month of Jan 2020",
//     amount: 5600,
//     createdAt: 5300
//   })
// );
// const exp1 = store.dispatch(addExpance({ ...expData[0] }));
// console.log("exp1 ", exp1.expence.id);
// store.dispatch(addExpance({ ...expData[1] }));

//store.dispatch(removeExpance(exp1.expence));

// store.dispatch(
//   editExpance({
//     id: exp1.expence.id,
//     update: {
//       note: "Net Bill",
//       Desc: "net bill for month of Jan 2020",
//       amount: 1100
//     }
//   })
// );

// note filter dispatch

// store.dispatch(noteFilter("net"));
// store.dispatch(sortByAmount());
