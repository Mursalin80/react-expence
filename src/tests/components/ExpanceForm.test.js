import { shallow } from "enzyme";
import React from "react";
import toJson from "enzyme-to-json";
import "../Enyzme.config";
import ExpanceForm from "../../components/ExpanceForm";
import { expData } from "../../redux/expenData";

test("should render expence form", () => {
  let wrap = toJson(shallow(<ExpanceForm />));
  expect(wrap).toMatchSnapshot();
});

test("should render expence form with data", () => {
  let wrap = toJson(shallow(<ExpanceForm expence={expData[2]} />));
  expect(wrap).toMatchSnapshot();
});

// simulate from summition
test("simulate from summition", () => {
  let wrap = shallow(<ExpanceForm onSubmit={() => {}} />);
  wrap.find("form").simulate("submit", {
    preventDefault: () => {}
  });
  expect(wrap.state("error").length).toBeGreaterThan(0);
  expect(toJson(wrap)).toMatchSnapshot();
});

test("simulate input note change", () => {
  let wrap = shallow(<ExpanceForm />);
  let value = "new note";

  wrap
    .find("input")
    .at(0)
    .simulate("change", {
      target: { value, name: "note" }
    });
  expect(wrap.state("note")).toBe(value);
  expect(toJson(wrap)).toMatchSnapshot();
});

test("simulate amount input change", () => {
  let wrap = shallow(<ExpanceForm />);
  let amount = "3000";
  wrap
    .find("input")
    .at(1)
    .simulate("change", {
      target: { value: amount, name: "amount" }
    });
  expect(wrap.state("amount")).toBe(amount);
});

test("onSubmit form handler", () => {
  let onSubmitSpy = jest.fn();
  let { note, description, amount, createdAt } = expData[2];
  let wrap = shallow(
    <ExpanceForm expence={expData[2]} onSubmit={onSubmitSpy} />
  );
  wrap.find("form").simulate("submit", {
    preventDefault: () => {}
  });
  expect(wrap.state("error")).toBe("");
  expect(onSubmitSpy).toHaveBeenLastCalledWith({
    note,
    description,
    amount,
    createdAt
  });
});
