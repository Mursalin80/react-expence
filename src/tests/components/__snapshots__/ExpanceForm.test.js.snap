// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render expence form 1`] = `
<div
  className="ui container"
>
  <form
    className="ui form"
    onSubmit={[Function]}
  >
    <input
      className="field"
      name="note"
      onChange={[Function]}
      placeholder="note"
      type="text"
      value=""
    />
    <input
      className="field"
      name="amount"
      onChange={[Function]}
      placeholder="amount"
      type="text"
      value={0}
    />
    <input
      className="field"
      name="description"
      onChange={[Function]}
      placeholder="description"
      type="text"
      value=""
    />
    <input
      className="field"
      name="createdAt"
      onChange={[Function]}
      placeholder="createdAt"
      type="text"
      value={0}
    />
    <button
      className="ui button"
    >
      Add Expence
    </button>
  </form>
</div>
`;

exports[`should render expence form with data 1`] = `
<div
  className="ui container"
>
  <form
    className="ui form"
    onSubmit={[Function]}
  >
    <input
      className="field"
      name="note"
      onChange={[Function]}
      placeholder="note"
      type="text"
      value="Milk bill"
    />
    <input
      className="field"
      name="amount"
      onChange={[Function]}
      placeholder="amount"
      type="text"
      value={5600}
    />
    <input
      className="field"
      name="description"
      onChange={[Function]}
      placeholder="description"
      type="text"
      value="Milk bill for month of Jan 2020"
    />
    <input
      className="field"
      name="createdAt"
      onChange={[Function]}
      placeholder="createdAt"
      type="text"
      value={5300}
    />
    <button
      className="ui button"
    >
      Add Expence
    </button>
  </form>
</div>
`;

exports[`simulate from summition 1`] = `
<div
  className="ui container"
>
  <form
    className="ui form"
    onSubmit={[Function]}
  >
    <input
      className="field"
      name="note"
      onChange={[Function]}
      placeholder="note"
      type="text"
      value=""
    />
    <input
      className="field"
      name="amount"
      onChange={[Function]}
      placeholder="amount"
      type="text"
      value={0}
    />
    <input
      className="field"
      name="description"
      onChange={[Function]}
      placeholder="description"
      type="text"
      value=""
    />
    <input
      className="field"
      name="createdAt"
      onChange={[Function]}
      placeholder="createdAt"
      type="text"
      value={0}
    />
    <button
      className="ui button"
    >
      Add Expence
    </button>
  </form>
</div>
`;

exports[`simulate input note change 1`] = `
<div
  className="ui container"
>
  <form
    className="ui form"
    onSubmit={[Function]}
  >
    <input
      className="field"
      name="note"
      onChange={[Function]}
      placeholder="note"
      type="text"
      value="new note"
    />
    <input
      className="field"
      name="amount"
      onChange={[Function]}
      placeholder="amount"
      type="text"
      value={0}
    />
    <input
      className="field"
      name="description"
      onChange={[Function]}
      placeholder="description"
      type="text"
      value=""
    />
    <input
      className="field"
      name="createdAt"
      onChange={[Function]}
      placeholder="createdAt"
      type="text"
      value={0}
    />
    <button
      className="ui button"
    >
      Add Expence
    </button>
  </form>
</div>
`;
